mono/metadata/threads.c

```diff
@@ -4584,6 +4607,69 @@ mono_thread_execute_interruption (void)
 	return exc;
 }
 
+static MonoExceptionHandle
+mono_thread_execute_interruption_handle (MonoError *error)
+{
+	MonoInternalThread *thread = mono_thread_internal_current ();
+	MonoExceptionHandle exc = MONO_HANDLE_NEW (MonoException, NULL);
+
+	LOCK_THREAD (thread);
+	gboolean unlock = TRUE;
+
+	/* MonoThread::interruption_requested can only be changed with atomics */
+	if (!mono_thread_clear_interruption_requested (thread))
+		goto exit;
+
+	MonoThreadObjectHandle sys_thread = MONO_HANDLE_NEW (MonoThreadObject, mono_thread_current ());
+
+	/* this will consume pending APC calls */
+#ifdef HOST_WIN32
+	MONO_ENTER_GC_SAFE;
+	mono_win32_wait_for_single_object_ex (GetCurrentThread (), 0, TRUE);
+	MONO_EXIT_GC_SAFE;
+#endif
+
+	/* Clear the interrupted flag of the thread so it can wait again */
+	mono_thread_info_clear_self_interrupt ();
+
+	/* If there's a pending exception and an AbortRequested, the pending exception takes precedence */
+	MONO_HANDLE_GET (exc, sys_thread, pending_exception);
+	if (!MONO_HANDLE_IS_NULL (exc)) {
+		MONO_HANDLE_SETRAW (sys_thread, pending_exception, NULL);
+		goto exit;
+	}
+
+	if (thread->state & ThreadState_AbortRequested) {
+		exc = MONO_HANDLE_NEW (MonoException, thread->abort_exc);
+		if (!MONO_HANDLE_IS_NULL (exc))
+			goto exit;
+		exc = mono_exception_new_thread_abort (error);
+		if (!MONO_HANDLE_IS_NULL (exc)) {
+			thread->abort_exc = MONO_HANDLE_RAW (exc);
```

Comment:
Isn't this missing a gc write barrier?  (the corresponding code in the version not using handles does a `MONO_OBJECT_SETREF` here)

Reply:
Maybe. It is within a lock, but all reads/writes would have to be for that to suffice and you are right it isn't "maintained in kind".